Next run configuration (#407)
openshift/openshift-origin-design: designs/virtualization/4.6/Next-run-configuration/Next-run-configuration.md

---
parent: Virtualization
version: 4.6
---

# Editing

When a user clicks to edit/ update one of the attributes (f.e. Flavor) it will open a modal with an alert notification letting the user know that the VM must be restarted in order to see the changes applied. The alert will persist until they restart the VM.
Alternatively they could clear the changes and keep what they had.
To avoid confusion they will be able to view the current settings in the VM Details page.

> ![user is editing values](img/EditOriginal.png)

Reviewer: If the user hasn't saved any Flavor changes yet, is the "Clear changes" button still visible in the alert? The very first time the modal is opened the Cancel button has the same functionality as Clear changes, so I wonder if they're both needed.

> ![user is editing values](img/EditOriginal.png)

Author: Thanks @andybraren. I totally agree we can have the alert as info instead of a warning one.
Yes, you are absolutely right about the button.

At this point the user can either just save the requested changes and postpone the restart or save and restart the VM.

## Restart modal

If the user chooses to save and restart the VM they will be asked again to ensure that and will get a modal including an inline notification that restarting may be disruptive to some logged in users and a notification that the new setting will be applied upon restart. The notification will include a confirmation checkbox so they would be aware of the effect that the restart action might cause and a link to the logged in users list at the bottom of the VM Details page.

![user chooses to resart the VM](img/RestartModal.png)

## VM Details page

In the VM Details page they will be notified that the changes they have made will get applied only after the VM will be restarted. 
Next to the ‘Flavor’ section we will add a link that will lead to the Flavor modal where they can view the pending changes.
Only in cases when users have made changes to the part of the YAML which are not exposed in the ui, we will add a link to ‘View pending changes in YAML editor’.

> ![VM details page](img/vm-details.png)

Author: Yes! This is a great opportunity to show the 'Edit Flavor' modal with the inline alert that lets the user clear that change.

> ![VM details page](img/vm-details.png)

Author: Thanks @matthewcarleton. I was looking for bullets example in PF4 but couldn't find any, but maybe it doesn't have to be a formal component?

In cases users made changes to the part of the YAML which are not exposed in the ui we’ll add the option to ‘View pending changes in the YAML editor.

![Users can view the changes in the YAML](img/ViewYAML-editor.png)

## List view

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Author: Thanks @lizsurette  Did you mean add an alert icon to the status column instead of the *?

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer 2: Exactly. Curious what others think of the idea too :) @andybraren @matthewcarleton

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Author: Thanks. I'll add both options so we can see better.

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer 3: ya I commented on the comment added by Andy. I think we could step back and look at the list view now that we are adding another "*". It might be out of scope of this effort but I like the idea of aligning with what we currently do and add the icon. In the future it could be nice to have an official status for Next Run and missing GA and drop the "*"

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer: By "alert icon" are we talking about the yellow warning triangle in the current design? If the inline alert within the Details page and modals changes to become a blue info type instead, a warning icon in the List may not align with what's shown there. This is a tricky balance. 😄

An option that might be more clear than the `*` could be `Changes pending` underneath the regular `Running` status. Kind of like what we do for operators:

![image](https://user-images.githubusercontent.com/9122899/82356698-f2dce980-99d1-11ea-9210-f6082690b41f.png)

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer 3: ya another reason why this is probably out of scope for this effort. I think there's room to improve the status column in general. 
@andybraren this is from the status card though right not in a list view? Be interesting to see how the list view would look with this. I'm all for some "Just in time" help :)

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Author: I agree it can definitely be clearer @andybraren but I wonder how it would look when not all the rows in the same height. I will create a mock-up and share it.

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer: It's from a list view yeah, sorry, I should've zoomed out. 🙂

![image](https://user-images.githubusercontent.com/9122899/82402652-f781bc00-9a2a-11ea-9fad-7898111c33c0.png)

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer 2: I really like the idea of using the subtext!

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Author: @andybraren @matthewcarleton @lizsurette WDYT of the 3rd option for the status suggested in the list view screen?

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer: Thinking more, one big benefit of treating un-restarted VMs as "degraded" with yellow warning statuses in List page is that they would/should appear in the Cluster Overview and (IIRC) Project Overview Inventory cards as well. That should make it even easier to notice VMs that need to be restarted. So yeah, good point @yfrimanm, sounds good to me! 👍 Should we mention that in this documentation?

And the substatus text lgtm!
![image](https://user-images.githubusercontent.com/9122899/82464045-9807c880-9a8b-11ea-8b3d-aff73e2797b8.png)

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer 3: nice! Ya I think we can create a follow up for the GA stuff to update the list view there too.

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Reviewer 2: This looks excellent to me!

> In the List view, we’ll show the Running* status. Clicking it will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.

Author: Thank you all @andybraren @matthewcarleton @lizsurette! Documentation & mockup are updated.

Users will be able to restart from here to save the step of clicking into the kebab menu.

> ![Popover indicating a restart is required from the list view](img/ListView.png)

Author: Thanks @andybraren! The idea of connecting from the popover to the VMs Details page is great and we definitely can use the 'X' button for canceling/closing.

You are right, I checked in PF and we don't have any example of a primary button in a popover so I changed it to be secondary.

> would the VM List's Filters dropdown support something like a "Pending changes" status to easily filter by ones that still need to be restarted?
@tjelinek@redhat.com will that be possible?

## Adding a ‘Clear changes’ option

Users will be able to view the pending changes from the ‘Edit’ modals and also have the option to clear the changes they have made by clicking the ‘Clear changes’ link.

![User edit and makes changes to values](img/Edit.png)

If they will choose to clear changes - that action will remove the changes they have made and the current settings will stay the same.

> ![Clear changes modal](img/ClearChanges2.png)

Reviewer: What does this modal look like if the user clicks the "Clear changes" inline alert on the Details page that could potentially list multiple pending changes for Flavor, Devices, etc.? Should each of those be listed in the modal's text here like this mockup does for flavor? Or would the modal not list them all and just be more general?

Author: In this case, I think the modal would include a list of all the pending changes with checkboxes so the user can select which ones they want to clear and which not.
WDYT? Can we support multiple selections now?

Author: @matthewcarleton what if they want to clear only 1 change out of many?

Reviewer 3: I'd say yes as a first pass on this. If we get feedback that it's really frustrating that they have to go to each one we could expand on it. I think we have to remember this is not a common thing a user will want to do. We shouldn't see a lot of cases where the user has made changes -> saved them -> and then regrets them. We just want to ensure they are covered if they find themselves in this situation.

Reviewer 4: A simple hack to clear some changes is to change them back to their previous configuration

## Disks and NICS examples

In Disks and NICs table view we’ll add ‘*pending changes’ link right beside the name so users will be able to see there are some changes that need to be applied.

> ![Disks page](img/Disks.png)

Reviewer: What are the chances that we could add a `Status` column for the Disks and NICs tables so that we could do something similar to the `Running*` status that's used in the VM List? 🙂 Maybe sometime in the future.

Author: > What are the chances that we could add a Status column for the Disks and NICs tables so that we could do something similar to the Running* status that's used in the VM List? 🙂 Maybe sometime in the future.

@tjelinek@redhat.com will that be possible?

Reviewer 3: I think we talked about that right @yfrimanm the only concern I saw was a status column that for the most part always has the same value. This is more of an edge case so adding a new column for it feels a bit heavy to me 🤷‍♂️

Author: Thanks for reminding me @matthewcarleton! yes, we agreed it would be too repetitive.

Reviewer 3: I wonder if we need a list view link here if we are going to have it in a notification as well. Maybe we can just rely on the notification for that. I made another comment about how that could look (disks -> Disk-1)

Author: +1, I think it makes sense.

![NICKs page](img/NICs.png)

The link will lead to the ‘Edit’ modal where the changes will be shown and users will have the option to clear them. If the user hasn't restarted the VM yet, we will add a  “Pending changes” label  beside the  items that have been changed.

> ![Edit NICK modal](img/EditNIC.png)

Reviewer: Should the same `*pending changes` seen here for the Network field also appear in the Edit Flavor modal example in this documentation? And since it's blue, does that mean it's clickable? Should it be black text instead?

I'm not sure that I've seen this pattern in PatternFly or the Web Console before. This may be worth bringing up with the group (or PatternFly) to see if anyone has encountered a need for this before, to see if making it a pattern/convention makes sense. From an accessibility perspective do you think this would just become part of the field label @matthewcarleton?

Author: I understand it's unclear because it looks like an odd link :) I guess we can give up of it is a link and make it black just to indicate there are pending changes.

---
parent: Virtualization
version: 4.6
---

# Editing

> When a user clicks to edit/ update one of the attributes (f.e. Flavor) it will open a modal with an alert notification letting the user know that the VM must be restarted in order to see the changes applied. The alert will persist until they restart the VM.

Author: +1

## Restart modal

If the user chooses to save and restart the VM they will be asked again to ensure that and will get into a modal including an inline notification that restarting may be disruptive to some logged in users and a notification that the new setting will be applied upon restart. The notification will include a confirmation checkbox so they would be aware of the effect that the restart action might cause and a link to the logged in users list at the bottom of the VM Details page.

![user chooses to resart the VM](img/RestartModal.png)

## VM Details page

In the VM Details page they will be notified that the changes they have made will get applied only after the VM will be restarted.
Next to the ‘Flavor’ section we will add a link that will lead to the Flavor modal where they can view the pending changes.
Only in cases when users have made changes to the part of the YAML which are not exposed in the ui, we will add a link to ‘View pending changes in YAML editor’.

![VM details page](img/vm-details.png)

In cases users made changes to the part of the YAML which are not exposed in the ui we’ll add the option to ‘View pending changes in the YAML editor.

> ![Users can view the changes in the YAML](img/ViewYAML-editor.png)

Reviewer 3: If we can detect changes were made in the YAML I think we should state that in the text
"View pending changes made in YAML" I think we need to know a bit more about the format of the text we could highlight from the YAML that was changed but it would be nice to be able to provide that here too. Otherwise we are just sending them to a YAML to try and find the change.  @jelkosz WDYT?

Author: +1

![Popover indicating a restart is required from the list view](img/ListView.png)

## Adding a ‘Clear changes’ option

Users will be able to view the pending changes from the ‘Edit’ modals and also have the option to clear the changes they have made by clicking the ‘Clear changes’ link.

Clicking 'Flavor' will open the 'Edit Flavor' modal with the inline alert that lets the user clear the change they made.

![Edit Flavor modal w inline alert that lets user clear change they made](img/Edit.png)

If they will choose to clear changes - that action will remove the changes they have made and the current settings will stay the same.

![Clear changes modal](img/ClearChanges2.png)

## Disks and NICS examples

In Disks and NICs table view we’ll add ‘*pending changes’ right beside the name so users will know there are some changes that need to be applied.

![Disks page](img/Disks.png)

![NICKs page](img/NICs.png)

In the ‘Edit’ modal the changes will be shown and users will have the option to clear them. If the user hasn't restarted the VM yet, we will add a  “Pending changes” label  beside the  items that have been changed.

> ![Edit NICK modal](img/EditNIC.png)

Reviewer 3: I wonder if we should have 
Network (pending changes)
instead of an asterisk since that denotes a required field otherwise.  WDYT?

Author: +1

Reviewer 3: nice!

Reviewer 3: @yfrimanm I think we should pull this. We can confirm with @irosenzw but I believe we are blocking users from opening these modals if changes are pending.

---
parent: Virtualization
version: 4.6
---
# Next run configuration

As a user makes changes to a running Virtual machine they will need to wait for the VM to restart to see the changes applied (the changes won’t be applied until the next start of the VM).

## Editing

When a user clicks to edit/ update one of the attributes (f.e. Flavor) it will open a modal with an alert notification letting the user know that the VM must be restarted in order to see the changes applied. The alert will persist until they restart the VM.
This alert will not appear until the user has made changes, so the user will then know that they will need to restart.

> In general, users will be able to clear the changes and keep what they had, but at the very first time the modal is opened in case the user hasn't saved any changes yet, the 'Clear changes' button will not be shown.

Reviewer 3: ```suggestion
In general, users will be able to clear the changes and keep what they had, but when the first time the modal is opened, in case the user hasn't saved any changes yet, the 'Clear changes' button will not be shown.
```

To avoid confusion users will be able to view the current settings in The VM Details page.

![user is editing values](img/EditOriginal.png)

At this point the user can either just save the requested changes and postpone the restart or save and restart the VM.

## Restart modal

If the user chooses to save and restart the VM they will be asked again to ensure that and will get into a modal including an inline notification that restarting may be disruptive to some logged in users and a notification that the new setting will be applied upon restart. The notification will include a confirmation checkbox so they would be aware of the effect that the restart action might cause and a link to the logged in users list at the bottom of the VM Details page.

> 

Reviewer 4: @yfrimanm @matthewcarleton What happens when the user click `no` when asking him the second time about the restart?
do we save the changes and return to the main view or we return to the first modal ?

Author: @irosenzw TMU we will save the changes and return to the VM details page.

![user chooses to resart the VM](img/RestartModal.png)

## VM Details page

In the VM Details page they will be notified that the changes they have made will get applied only after the VM will be restarted.
Next to the ‘Flavor’ section we will add a link that will lead to the Flavor modal where they can view the pending changes.
Only in cases when users have made changes to the part of the YAML which are not exposed in the ui, we will add a link to ‘View pending changes in YAML editor’.

> ![VM details page](img/vm-details.png)

Reviewer 3: just thinking about this high level notification a bit more. So for changes that are on other tabs (Disks, Nics, Environment) we could do 
- Flavor
- Disks > Disk-1
- Network interfaces -> nic-1
- Environment
Then we show this same notification on every tab and we don't need to show separate one like you have currently. Clicking each takes the user to the tab and opens the modal or just shows the environment.
For Environment I think we'll still need a separate inline notification about changes there and the ability to clear them

Author: Thanks @irosenzw I'll add this clarification to the doc.

Reviewer 5: IMO, we shouldn't allow changes to Disks/CD-ROMs. This can cause the underlying PVC which is the VMI using to be deleted.

![user is editing values](img/EditOriginal.png)

At this point the user can either just save the requested changes and postpone the restart or save and restart the VM.

## Restart modal

If the user chooses to save and restart the VM they will be asked again to ensure that and will get into a modal including an inline notification that restarting may be disruptive to some logged in users and a notification that the new setting will be applied upon restart. The notification will include a confirmation checkbox so they would be aware of the effect that the restart action might cause and a link to the logged in users list at the bottom of the VM Details page.

![user chooses to resart the VM](img/RestartModal.png)

## VM Details page

In the VM Details page they will be notified that the changes they have made will get applied only after the VM will be restarted.
Next to the ‘Flavor’ section we will add a link that will lead to the Flavor modal where they can view the pending changes.
Only in cases when users have made changes to the part of the YAML which are not exposed in the ui, we will add a link to ‘View pending changes in YAML editor’.

![VM details page](img/vm-details.png)

> In case there are multiple changes that are on different tabs, we will show a list of them and the same notification will be shown on every tab. Clicking each will take the user to the relevant tab and will open the modal.

Author: We already have an inline warning alert at the top of the VM Details page AND 'View pending changes' link beside each change.

Clicking the 'Running' status will open a popover explaining that this VM has some pending changes that will apply once it will be restarted.
Users will be able to restart from the popover to save the step of clicking into the kebab menu.
The popover will also let the user know where to see the pending changes by clicking 'View details' which will take them to the VMs Details page so they can see the 'Restart is required' inline alert at the top.

![Popover indicating a restart is required from the list view](img/ListView.png)

## Adding a ‘Clear changes’ option

Users will be able to view the pending changes from the ‘Edit’ modals and also have the option to clear the changes they have made by clicking the ‘Clear changes’ link.

Clicking 'Flavor' will open the 'Edit Flavor' modal with the inline alert that lets the user clear the change they made.

> We will add a (Pending changes) label beside the items that have been changed.

Reviewer 4: Maybe instead of labels we can paint the pencil icon in yellow or add a warning sign beside it and have a popup that shows the 'pending changes label'.

Another idea is to have a specific (unique) icon for pending changes warnings and we can put this icon beside each item that has pending changes. WDYT ?

Author: I thought about the colored pencil as well but it's not accessible for users with visual limitations. We also don't surface pop up over a pop up.

We use PF4 framework and currently we don't have such an icon, but it is something we might recheck in the future.

---
parent: Virtualization
version: 4.6
---
# Next run configuration

As a user makes changes to a running Virtual machine they will need to wait for the VM to restart to see the changes applied (the changes won’t be applied until the next start of the VM).

## Editing

When a user clicks to edit/ update one of the attributes (f.e. Flavor) it will open a modal with an alert notification letting the user know that the VM must be restarted in order to see the changes applied. The alert will persist until they restart the VM.
This alert will not appear until the user has made changes, so the user will then know that they will need to restart.

In general, users will be able to clear the changes and keep what they had, but at the very first time the modal is opened in case the user hasn't saved any changes yet, the 'Clear changes' button will not be shown.

> To avoid confusion users will be able to view the current settings in The VM Details page.

Reviewer 5: Current settings of VM ? or VMI ?

Reviewer 5: linking a discussion in implementation: https://github.com/openshift/console/pull/5633/files#r433397144

## VM Details page

In the VM Details page they will be notified that the changes they have made will get applied only after the VM will be restarted.
In the ‘Flavor’ section, beside the Edit icon, we will add (pending changes) that will lead to the Edit Flavor modal where they can view the pending changes.
Only in cases when users have made changes to the part of the YAML which are not exposed in the ui, we will add a link to ‘View pending changes in YAML editor’.

![VM details page](img/vm-details.png)

In case there are multiple changes that are on different tabs, we will show a list of them and the same notification will be shown on every tab. Clicking each will take the user to the relevant tab and will open the modal.
An environment change can also be presented as a Disk change (if a user adds an environment key-value it is being added as a disk), so a change in the environment tab should be reflected in both 'Disks' and 'Environment' tabs and include a separate inline notification about the changes with the ability to clear them.

For changes that apply across all tabs, we'll place the notification above the tabs, so users will always be able to see it.

> ![Notification for changes on multiple tabs](img/Notification-changes-on-multiple-tabs.png)

Reviewer 3: talking with @itsptk about this a bit more. I think having this alert always above the tabs makes the most sense. Even if there is only one issue that it's addressing. It seems more sensible. 
The only change required would be to list them out with the same format
details > flavor

Reviewer 6: Agreed, it seems like even if only one change was made on a particular tab, you'd still be interested in knowing that there are pending changes even if you were now looking at a different tab.

Author: +1
